Lib/pathlib.py

```diff
@@ -1385,6 +1385,132 @@ def expanduser(self):
+        optional arg 'on_error' is specified, it should be a function; it
+        will be called with one argument, an OSError instance.  It can
+        report the error to continue with the walk, or raise the exception
+        to abort the walk.  Note that the filename is available as the
+        filename attribute of the exception object.
+
+        By default, Path.walk does not follow symbolic links to subdirectories
+        on systems that support them.  In order to get this functionality, set
+        the optional argument 'follow_links' to true.
+
+        Caution: if self is a relative Path, don't change the
+        current working directory between resumptions of walk. walk never
+        changes the current directory, and assumes that the client doesn't
+        either.
+
+        Example:
+
+        from pathlib import Path
+        for root, dirs, files in Path('python/Lib/email'):
+            print(root, "consumes", end="")
+            print(sum((root / file).stat().st_size for file in files), end="")
+            print("bytes in", len(files), "non-directory files")
+            if 'CVS' in dirs:
+                dirs.remove('CVS')  # don't visit CVS directories
+        """
+        sys.audit("pathlib.Path.walk", self, on_error, follow_links)
+        return self._walk(True, on_error, follow_links)
+
+    def walk_bottom_up(self, on_error=None, follow_links=False):
+        """Generate a bottom up directory tree from this directory
+
+        The return type and arguments are identical to Path.walk.
+        However, the caller cannot modify the child_directory_names to prune
+        the search or to impose a specific order of visiting because the list
+        of directories to visit is calculated before we yield it.
+        """
+        sys.audit("pathlib.Path.walk_bottom_up", self, on_error, follow_links)
+        return self._walk(False, on_error, follow_links)
+
+    def _walk(self, topdown, on_error, follow_links):
+        dirs = []
+        nondirs = []
+        walk_dirs = []
+
+        # We may not have read permission for self, in which case we can't
+        # get a list of the files the directory contains. os.walk
+        # always suppressed the exception then, rather than blow up for a
+        # minor reason when (say) a thousand readable directories are still
+        # left to visit. That logic is copied here.
+        try:
+            scandir_it = self._scandir()
+        except OSError as error:
+            if on_error is not None:
+                on_error(error)
+            return
+
+        with scandir_it:
+            while True:
+                try:
+                    entry = next(scandir_it, ...)
+                    if entry is ...:
+                        break
+                except OSError as error:
+                    if on_error is not None:
+                        on_error(error)
+                    return
+
+                try:
+                    is_dir = entry.is_dir()
```

Comment:
```suggestion
                    is_dir = entry.is_dir(follow_symlinks=follow_symlinks)
```

Wouldn't this allow us to do away with all the remaining symlink code?

Reply:
No. This would help us with walk_dirs behavior but it will break the dirnames behavior because symlinks will be listed as nondirs, even if they point to directories. And followlinks argument only prevents/allows for **descent** into those directories. It shouldn't affect how we list directories.

Reply:
> symlinks will be listed as nondirs, even if they point to directories

Symlinks that point to directories would be listed in `filenames` when `follow_symlinks=False`, and in `dirnames` when `follow_symlinks=True`. What's the problem there?

Reply:
The problem is that they are still directories, not files. At least in the original walk. Even if we do not descend into them.

And I believe this behavior makes a lot of sense. **follow** symlinks instructs whether walk **follows** them, not **recognizes** them as directories.

Reply:
Gotcha. I'd like opinions from others on this as personally I find that pretty odd. If I set `follow_symlinks=False` I wouldn't expect symlinks to show up in `dirnames`.